Assets/Scripts/Foliage/SpawnFoliage.cs

```diff
@@ -1,161 +1,247 @@
-    [SerializeField] private float spawningAngleLimit;
-    [SerializeField] private int stoneLimit;
-    [SerializeField] private int bushLimit;
-    [SerializeField] private int treeLimit;
-    [SerializeField] private int seed;
-    [SerializeField] private float treeLine;
-
-    [SerializeField] private Object[] trees = new Object[4];
-    [SerializeField] private Object[] bushes = new Object[2];
-    //[SerializeField] private Object[] grass = new Object[2];
-    [SerializeField] private Object[] stones = new Object[2];
-
-    // Start is called before the first frame update
-    void Start()
-    {
-        Random.InitState(seed);
+{
+    const int prefabLimit = 10000;
 
-        plantStones();
-        plantBushes();
-        plantTrees();
-    }
+    [Header("Trees")]
+    [SerializeField] private int treeLimit = 1000;
+    [SerializeField] private float treeLine = 25;
+    [SerializeField] private float treeAngleLimit = 30;
+    [SerializeField] private GameObject[] treePrefabs = new GameObject[4];
 
-    /// <summary>
-    /// Plants trees
-    /// </summary>
-    private void plantTrees()
+    [Header("Bushes")]
+    [SerializeField] private int bushLimit = 5000;
+    [SerializeField] private float bushLine = 35;
+    [SerializeField] private float bushAngleLimit = 30;
+    [SerializeField] private GameObject[] bushPrefab = new GameObject[4];
+
+    [Header("Rocks")]
+    [SerializeField] private int stoneLimit = 400;
+    [SerializeField] private float stoneLine = 36;
+    [SerializeField] private float stoneAngleLimit = 30;
+    [SerializeField] private GameObject[] stonePrefab = new GameObject[4];
+
+    [Header("Misc")]
+    [SerializeField] private int seed = 0;
+    [SerializeField] private bool DEBUG = false;
+
+
+    private int treeIndex = 0;
+    private int treeSpawnIndex = 0;
+    private Vector3[] treePositions = new Vector3[prefabLimit];
+
+    private int bushIndex = 0;
+    private int bushSpawnIndex = 0;
+    private Vector3[] bushPositions = new Vector3[prefabLimit];
+
+    private int stoneIndex = 0;
+    private int stoneSpawnIndex = 0;
+    private Vector3[] stonePositions = new Vector3[prefabLimit];
+
+
+    // Private members
+    private GameObject foliageHandler;
+
+    private Vector3 planetCenter;
+    private Planet planet;
+    private float planetRadius;
+    private float waterLevel;
+
+    void Update()
     {
-        Vector3 position;
-        RaycastHit hit;
 
-        // Loops thought all tree locations
-        for (int i = 0; i < treeLimit; i++)
+        if(planet != null && foliageHandler != null)
         {
-            // Creates a ball around the planet which shoots rays to check if it's legal to spawn foliage
-            position = Random.onUnitSphere * rayBallRadius;
-
-            if (Physics.Raycast(position, -position, out hit, rayLenght))
+            GameObject[] cameras = GameObject.FindGameObjectsWithTag("MainCamera");
+            for(int i = 0; i < cameras.Length; i++)
```

Comment:
Iterating through all cameras means that the foliage may be destroyed by one camera and created by another. Either use only one main camera or check if all cameras are out of range before despawning.

Reply:
The "Find" method is very slow as it iterates through all game objects in the hierarchy. This is not good to have in a "Update" method which runs every frame. I would remove it and find some other way to save performance

Reply:
Made it so that only one camera can alter the spawning of foliage.
Moved "Find" to only find the player at initialization.